[Filebeat][New Input] Http Input

P1llus opened the pull request:
## What does this PR do?
This filebeat input configures a HTTP port listener, accepting JSON formatted POST requests, which again is formatted into a event, initially the event is created with the "json." prefix and expects the ingest pipeline to mutate the event during ingestion.

The initial set of features is based on the Logstash input plugin, but implemented differently: https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http.html

## Why is it important?
This idea is based on a few different scenarios:
- The user already has a large beats installation and no Logstash, and do not want to install Logstash solely for a single feature.
- HTTP Input allows applications to be directly integrated with Elastic, without needing connectivity to Elasticsearch directly (or Logstash).
- Allows us to integrate and create modules for any product that supports HTTP POST events like SOAR, cloud applications, ticketing systems etc etc.

## Features currently implemented
HTTP Basic Auth On/Off
HTTP/HTTPS configurable
Listening interface and port configurable
Response code on success configurable
Response body on success configurable
Response header on success configurable
Proper HTTP codes on both success and error responses
Message prefix configurable
URL to post to is configurable 
SSL path to cert, key and CA is configurable.

## TODO
- [x] Clean up code
- [x] Import configuration options from common
- [x] Write documentation
- [x] Add TLS mutual auth support
- [x] Config validation
- [x] Tests


## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
## :broken_heart: Build Failed 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_18298.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #18298 opened]
* Start Time: 2020-05-06T11:29:28.447+0000

* Duration: 16 min 48 sec (947846)


* Commit: c94b6b91911240cb50e5b94cdd0c9154dba8119a




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: Make check
    * Description: make check
    * Result: FAILURE
    
    * Duration: 9 min 4 sec<
    
    * Start Time: 2020-05-06T11:35:42.076+0000
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-05-06T11:44:50.880Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:50.885Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:50.891Z] Stage "Winlogbeat Windows x-pack" skipped due to earlier failure(s)
[2020-05-06T11:44:50.900Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:50.905Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:50.910Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-06T11:44:50.915Z] Stage "Kubernetes" skipped due to earlier failure(s)
[2020-05-06T11:44:51.143Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.153Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-06T11:44:51.162Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.172Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-05-06T11:44:51.181Z] Stage "Packetbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.190Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.200Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.216Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.230Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:51.243Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-06T11:44:51.947Z] Failed in branch Elastic Agent x-pack
[2020-05-06T11:44:51.955Z] Failed in branch Elastic Agent x-pack Windows
[2020-05-06T11:44:51.963Z] Failed in branch Elastic Agent Mac OS X
[2020-05-06T11:44:51.971Z] Failed in branch Filebeat oss
[2020-05-06T11:44:51.981Z] Failed in branch Filebeat x-pack
[2020-05-06T11:44:51.989Z] Failed in branch Filebeat Mac OS X
[2020-05-06T11:44:51.998Z] Failed in branch Filebeat Windows
[2020-05-06T11:44:52.007Z] Failed in branch Auditbeat x-pack
[2020-05-06T11:44:52.015Z] Failed in branch Libbeat x-pack
[2020-05-06T11:44:52.023Z] Failed in branch Metricbeat OSS Unit tests
[2020-05-06T11:44:52.033Z] Failed in branch Metricbeat OSS Integration tests
[2020-05-06T11:44:52.041Z] Failed in branch Metricbeat Python integration tests
[2020-05-06T11:44:52.049Z] Failed in branch Metricbeat crosscompile
[2020-05-06T11:44:52.057Z] Failed in branch Metricbeat Mac OS X
[2020-05-06T11:44:52.065Z] Failed in branch Metricbeat Windows
[2020-05-06T11:44:52.073Z] Failed in branch Winlogbeat Windows x-pack
[2020-05-06T11:44:52.081Z] Failed in branch Kubernetes
[2020-05-06T11:44:52.646Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:52.655Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-06T11:44:52.664Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:52.672Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-05-06T11:44:52.680Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:52.687Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:52.695Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-06T11:44:52.983Z] Failed in branch Packetbeat
[2020-05-06T11:44:52.991Z] Failed in branch dockerlogbeat
[2020-05-06T11:44:52.999Z] Failed in branch Journalbeat
[2020-05-06T11:44:53.413Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:53.422Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-06T11:44:53.431Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:53.439Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-06T11:44:53.447Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-06T11:44:53.541Z] Failed in branch Metricbeat x-pack
[2020-05-06T11:44:53.550Z] Failed in branch Winlogbeat
[2020-05-06T11:44:53.945Z] Failed in branch Heartbeat
[2020-05-06T11:44:53.954Z] Failed in branch Libbeat
[2020-05-06T11:44:53.964Z] Failed in branch Functionbeat
[2020-05-06T11:44:53.965Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-06T11:44:53.974Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-06T11:44:54.216Z] Failed in branch Auditbeat oss
[2020-05-06T11:44:54.224Z] Failed in branch Generators
[2020-05-06T11:44:55.398Z] Running in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-06T11:44:55.730Z] + find . -type f -name TEST*.xml -path */build/* -delete
[2020-05-06T11:44:55.753Z] Running in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats/Lint
[2020-05-06T11:44:56.264Z] + cat
[2020-05-06T11:44:56.264Z] + /usr/local/bin/runbld ./runbld-script
[2020-05-06T11:44:56.264Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-05-06T11:45:02.844Z] runbld>>> runbld started
[2020-05-06T11:45:02.844Z] runbld>>> 1.6.11/a66728ff8f4356963772e6e6d2069392fa06acbe
[2020-05-06T11:45:03.786Z] runbld>>> The following profiles matched the job 'Beats/beats-beats-mbp/PR-18298' in order of occurrence in the config (last value wins).
[2020-05-06T11:45:05.167Z] runbld>>> Debug logging enabled.
[2020-05-06T11:45:05.167Z] runbld>>> Storing result
[2020-05-06T11:45:05.428Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-05-06T11:45:05.428Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1587637540455/t/20200506114505-0F9CC11A
[2020-05-06T11:45:05.428Z] runbld>>> Adding system facts.
[2020-05-06T11:45:05.999Z] runbld>>> Adding vcs info for the latest commit:  c94b6b91911240cb50e5b94cdd0c9154dba8119a
[2020-05-06T11:45:06.570Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-05-06T11:45:06.570Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-05-06T11:45:06.570Z] + echo 'Processing JUnit reports with runbld...'
[2020-05-06T11:45:06.570Z] Processing JUnit reports with runbld...
[2020-05-06T11:45:06.831Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-05-06T11:45:06.831Z] runbld>>> DURATION: 8ms
[2020-05-06T11:45:06.831Z] runbld>>> STDOUT: 40 bytes
[2020-05-06T11:45:06.831Z] runbld>>> STDERR: 49 bytes
[2020-05-06T11:45:06.831Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-05-06T11:45:06.832Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-06T11:45:08.217Z] runbld>>> Storing build metadata: 
[2020-05-06T11:45:08.217Z] runbld>>> Adding test report.
[2020-05-06T11:45:08.217Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-06T11:45:08.791Z] runbld>>> Found 0 test output files
[2020-05-06T11:45:08.791Z] runbld>>> Test output logs contained: Errors: 0 Failures: 0 Tests: 0 Skipped: 0
[2020-05-06T11:45:09.062Z] runbld>>> Storing result
[2020-05-06T11:45:09.062Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-05-06T11:45:09.062Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1587637540455/t/20200506114505-0F9CC11A
[2020-05-06T11:45:09.063Z] runbld>>> Email notification disabled by environment variable.
[2020-05-06T11:45:09.063Z] runbld>>> Slack notification disabled by environment variable.
[2020-05-06T11:45:15.293Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298
[2020-05-06T11:45:15.480Z] [INFO] getVaultSecret: Getting secrets
[2020-05-06T11:45:15.565Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-05-06T11:45:16.564Z] + chmod 755 generate-build-data.sh
[2020-05-06T11:45:16.564Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/runs/1 FAILURE 947846
[2020-05-06T11:45:17.115Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/runs/1/steps/?limit=10000 -o steps-info.json
[2020-05-06T11:45:17.666Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>

andrewkroh commented on x-pack/filebeat/input/httpinput/config.go:
```diff
@@ -0,0 +1,42 @@
+
+// Config contains information about httpjson configuration
+type config struct {
+	UseSSL          bool   `config:"ssl"`
```
All of the TLS options can be defined by reusing `tlscommon`.


```suggestion
	TLS  *tlscommon.ServerConfig `config:"ssl"`
```

P1llus replied:
That's a good point @andrewkroh . I looked at it briefly, as I saw some other inputs used it, and it was a bit unclear which of my options was covered by it and not. If I include tlscommon, does it only cover the settings in config.go located in that folder?
Ref:
```
type Config struct {
	Enabled          *bool                   `config:"enabled" yaml:"enabled,omitempty"`
	VerificationMode TLSVerificationMode     `config:"verification_mode" yaml:"verification_mode"` // one of 'none', 'full'
	Versions         []TLSVersion            `config:"supported_protocols" yaml:"supported_protocols,omitempty"`
	CipherSuites     []tlsCipherSuite        `config:"cipher_suites" yaml:"cipher_suites,omitempty"`
	CAs              []string                `config:"certificate_authorities" yaml:"certificate_authorities,omitempty"`
	Certificate      CertificateConfig       `config:",inline" yaml:",inline"`
	CurveTypes       []tlsCurveType          `config:"curve_types" yaml:"curve_types,omitempty"`
	Renegotiation    tlsRenegotiationSupport `config:"renegotiation" yaml:"renegotiation"`
	CASha256         []string                `config:"ca_sha256" yaml:"ca_sha256,omitempty"`
}
```

andrewkroh replied:
Correct, that's what it covers. You need to provide all the other non-TLS options. By reusing it users get a consistent set of options across the stack.

https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/libbeat/common/transport/tlscommon/server_config.go#L28-L38

P1llus replied:
Okay thanks will add it in for the next version then! Always appreciate feedback, will wait a bit for more comments if there is any and implement all of them in one go :)

kvch commented on x-pack/filebeat/input/httpinput/config.go:
```diff
@@ -0,0 +1,42 @@
+	BasicAuth       bool   `config:"basic_auth"`
+	Username        string `config:"username"`
+	Password        string `config:"password"`
+	ResponseCode    int    `config:"response_code"`
```
Do you mind adding the following requirements to `ResponseCode`?
```golang
	ResponseCode    int    `config:"response_code" validate:"positive"`
```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+
+	in := &HttpInput{
+		config:       conf,
+		log:          logp.NewLogger("httpinput"),
```
You could use `inputName` here, so when the input name is changed, we don't need to change here as well.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+		workerCancel: workerCancel,
+	}
+
+	in.log.Info("Initialized httpinput input.")
```
Use `inputName` here, too.

kvch replied:
Maybe add a bit more context here, e.g. the configured `listen_port` and `listen_address`.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	in.httpMux = http.NewServeMux()
+	in.httpMux.HandleFunc(in.config.URL, in.apiResponse)
+
+	if in.config.UseSSL == true {
```
Please rewrite the following part as `in.config.UseSSL == true` is redundant.
```golang
	if in.config.UseSSL {
		in.httpServer = &http.Server{Addr: address.String(), Handler: in.httpMux}
		return in.httpServer.ListenAndServeTLS(in.config.SSLCertificate, in.config.SSLKey)
	}
	in.httpServer = &http.Server{Addr: address.String(), Handler: in.httpMux}
	return in.httpServer.ListenAndServe()
```

Also, please remove the following line because it is dead code. `in.config.UseSSL` is either `true` or `false`. There is no third value for `bool`. So this line is never reached.
```golang
	return errors.New("SSL settings missing")
```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+
+func (in *HttpInput) createServer() error {
+	// Merge listening address and port
+	var address strings.Builder
```
I think it is simpler to just use `+` to concat strings. There is no need to use `strings.Builder`:
```golang
address := in.config.ListenAddress + ":" + in.config.ListenPort
```
BTW I've seen some benchmarks which state that using `+` is the quickest way to concat string in Go. (Ref: https://dev.to/pmalhaire/concatenate-strings-in-golang-a-quick-benchmark-4ahh)

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	var status uint
+
+	// Check auth settings and credentials
+	if in.config.BasicAuth == true {
```
```suggestion
	if in.config.BasicAuth {
```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+
+	// Write full []byte to string
+	body, err := ioutil.ReadAll(in.httpRequest.Body)
+
```
Please remove this empty line.

```suggestion

```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	// Write full []byte to string
+	body, err := ioutil.ReadAll(in.httpRequest.Body)
+
+	// If body cannot be read
```
This comment is unnecessary.
```suggestion

```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	// Declare interface for request body
+	objmap := make(map[string]interface{})
+
+	err = json.Unmarshal(body, &objmap)
```
Same as the above (no empty newline and unnecessary comment).

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+}
+
+func (in *HttpInput) validateMethod() (uint, string) {
+	// Ensure HTTP method is POST
```
This comment seems to me like a "header" comment which should go above the function declaration. E.g.
```golang
// validateMethod ensures that the HTTP method is POST
func (in *HttpInput) validateMethod() (uint, string) {
```

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("httpinput worker has started.")
```
In both log statements it would be nice to get more context, too. E.g. `listen_host` and `listen_port`.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	workerOnce   sync.Once               // Guarantees that the worker goroutine is only started once.
+	workerWg     sync.WaitGroup          // Waits on worker goroutine.
+	httpServer   *http.Server            // The currently running HTTP instance
+	httpMux      *http.ServeMux          // Current HTTP Handler
```
I don't think this attribute is needed. One can reach this from `in.httpServer.Handler`. Besides I don't see it used anywhere in the code.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	defer cancel()
+
+	// Initialize the HTTP server
+	err = in.createServer()
```
This name is a bit misleading. It not only creates/initialize a server but starts it, too.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *HttpInput) Run() {
```
To me it seems that this function is a bit too complicated with all of the context and for loops. I think you can simplify it if you took a look at the HTTPServer implementation in Metricbeat: https://github.com/elastic/beats/blob/master/metricbeat/helper/server/http/http.go

The functionality provided by the HTTP module of Metricbeat.

kvch commented on x-pack/filebeat/input/httpinput/input.go:
```diff
@@ -0,0 +1,329 @@
+	ok := in.outlet.OnEvent(beat.Event{
+		Timestamp: time.Now().UTC(),
+		Fields: common.MapStr{
+			"message":        "testing",
```
Is this a leftover from testing?

P1llus commented:
@andrewkroh : Moved to tlscommon, though a small issue described below

@kvch : Added in your changes and moved http server into its own file to keep it a bit more simplified and more like the Metricbeat one, though not 100% the same.

There is one issue with tlscommon, and that is the fact that it does not allow to just spin up a HTTP server without client authentication, as it currently enforces it when certificate and key configuration option is there.

Currently I have bypassed this by creating another config option under ssl, called `ssl.client_auth`, if that is set to false, it is disabled during `createServer()`.

There is a few issues with this, the main one is that when I will create the documentation for this, I would not be able to include the documentation unless I can override it somehow. It should be possible for someone to run HTTPS without having to enable client authentication.

What would be the best way to approach this?

kvch commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,290 @@
+// will ever start the worker.
+func (in *HttpEndpoint) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
```
You have to call `in.workerWg.Done()` from the function `in.run`. The goroutine you start is executed "in paralel", it return immediately. Thus, `in.workerWg.Done()` is called right after the start. To wait until the execution of `in.run` is done, you need to rewrite the code.

```golang
func (in *HttpEndpoint) Run() {
	in.workerOnce.Do(func() {
		in.workerWg.Add(1)
		go in.run()
	})
}

func (in *HttpEndpoint) run() {
	defer in.workerWg.Done()
	defer in.log.Infof("%v worker has stopped.", inputName)
	err := in.server.Start()
	if err != nil {
		return
	}

	return
}
```

kvch commented on x-pack/filebeat/input/http_endpoint/httpserver.go:
```diff
@@ -0,0 +1,81 @@
+		ctx:  ctx,
+		stop: cancel,
+	}
+	fmt.Printf("%+v\n", defaultConfig)
```
Leftover from debugging?

kvch commented on x-pack/filebeat/input/http_endpoint/httpserver.go:
```diff
@@ -0,0 +1,81 @@
+func (h *HttpServer) Start() error {
+	go func() {
+		if h.server.TLSConfig != nil {
+			logp.Info("Starting HTTPS server on %s", h.server.Addr)
```
Please add a logger to `HttpServer`, so you can avoid using `logp.Info`. We are trying to get rid of these global functions ATM.

kvch commented:
> @andrewkroh : Moved to tlscommon, though a small issue described below
> 
> @kvch : Added in your changes and moved http server into its own file to keep it a bit more simplified and more like the Metricbeat one, though not 100% the same.
> 
> There is one issue with tlscommon, and that is the fact that it does not allow to just spin up a HTTP server without client authentication, as it currently enforces it when certificate and key configuration option is there.
> 
> Currently I have bypassed this by creating another config option under ssl, called `ssl.client_auth`, if that is set to false, it is disabled during `createServer()`.
> 
> There is a few issues with this, the main one is that when I will create the documentation for this, I would not be able to include the documentation unless I can override it somehow. It should be possible for someone to run HTTPS without having to enable client authentication.
> 
> What would be the best way to approach this?

I am not sure I understand the problem. I don't see where client authentication gets configured by `tlscommon` in your case. 

You can configure [ClientAuthentication](https://github.com/elastic/beats/blob/master/libbeat/common/transport/tlscommon/server_config.go#L37) in `tlscommon.ServerConfig`. Only `LoadTLSConfig` enforces client authentication. But you are using `LoadTLSServerConfig`.

P1llus commented:
> > @andrewkroh : Moved to tlscommon, though a small issue described below
> > @kvch : Added in your changes and moved http server into its own file to keep it a bit more simplified and more like the Metricbeat one, though not 100% the same.
> > There is one issue with tlscommon, and that is the fact that it does not allow to just spin up a HTTP server without client authentication, as it currently enforces it when certificate and key configuration option is there.
> > Currently I have bypassed this by creating another config option under ssl, called `ssl.client_auth`, if that is set to false, it is disabled during `createServer()`.
> > There is a few issues with this, the main one is that when I will create the documentation for this, I would not be able to include the documentation unless I can override it somehow. It should be possible for someone to run HTTPS without having to enable client authentication.
> > What would be the best way to approach this?
> 
> I am not sure I understand the problem. I don't see where client authentication gets configured by `tlscommon` in your case.
> 
> You can configure [ClientAuthentication](https://github.com/elastic/beats/blob/master/libbeat/common/transport/tlscommon/server_config.go#L37) in `tlscommon.ServerConfig`. Only `LoadTLSConfig` enforces client authentication. But you are using `LoadTLSServerConfig`.

After adding a CA, certificate and key, it still ended up enforcing client authentication. Trying to do a common HTTPS request responded with missing certificates for client authentication.

The other issue would be in regards to the documentation, since the TLS/SSL documentation states that client auth will be enforced if a certificate is chosen, I could always write my own instead of linking to the common page, whichever you prefer :)

kvch commented on x-pack/filebeat/input/http_endpoint/httpserver.go:
```diff
@@ -0,0 +1,81 @@
+
+	http.HandleFunc(in.config.URL, in.apiResponse)
+	tlsConfig, err := tlscommon.LoadTLSServerConfig(in.config.TLS)
+
```
```suggestion

```

kvch commented on x-pack/filebeat/input/http_endpoint/httpserver.go:
```diff
@@ -0,0 +1,81 @@
+	server *http.Server
+	ctx    context.Context
+	stop   context.CancelFunc
+	done   chan struct{}
```
Is this channel necessary? I don't see it used anywhere.

kvch commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,290 @@
+
+func (in *HttpEndpoint) run() {
+	defer in.log.Infof("%v worker has stopped.", inputName)
+	err := in.server.Start()
```
You can ignore this error, as it is always nil.
```
func (in *HttpEndpoint) run() {
	defer in.log.Infof("%v worker has stopped.", inputName)
	in.server.Start()
}
```

kvch replied:
Also, I have a feeling this `run` returns immediately, as it is starting a goroutine. It is probably not necessary to use a goroutine there. I am not 100% sure as I haven't run the code.

P1llus replied:
If func (h *HttpServer) Start() is not a goroutine then the http server does not close when doing CTRL+C, and it gets stuck shutting down. But I removed the return for now.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/httpserver.go:
```diff
@@ -0,0 +1,71 @@
+func (h *HttpServer) Stop() {
+	h.log.Info("Stopping HTTP server")
+	h.stop()
+	h.server.Shutdown(h.ctx)
```
Should this take a context with some sort of timeout? Otherwise it could block for a very long time if somehow a connection is still active.

P1llus replied:
I have modified my context from TODO to background with a timeout, hopefully this is what you meant?

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,277 @@
+
+// Validates the incoming request and creates a new event upon success
+func (in *HttpEndpoint) sendEvent(w http.ResponseWriter, r *http.Request) (uint, string) {
+	status, err := in.validateRequest(w, r)
```
Just a thought, but I think that all of this validation code might be more easily expressed in some sort of middleware pattern.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,277 @@
+	return ""
+}
+
+func (in *HttpEndpoint) sendResponse(w http.ResponseWriter, h uint, b string) {
```
shouldn't this method set the configured `ResponseHeaders`? That said, what keeps someone from overriding the `Content-Type` header or passing off some other form of response in the configured `ResponseBody`?

P1llus replied:
sendResponse is also for sending error messages, the configured ResponseHeaders should only be configured upon successful request, so headers and body should be set in `apiResponse()`.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,277 @@
+		return
+	}
+
+	in.sendResponse(w, http.StatusOK, in.config.ResponseBody)
```
is this supposed to be `in.config.ResponseCode` rather than `http.StatusOK`?

P1llus replied:
That is correct. This is something that I should have picked up. Initially it was put there since I found no good way to convert an int/uint to a StatusType, still looking into that.

P1llus commented:
@andrewstucki 
Thanks for all the comments, greatly appreciated! :)

My last commit added a few changes:

- I did not want to make any complex middleware implementation or import any extra libraries for that purpose, currently validate is a single middleware function, which still calls all other validation functions. Is this okay or would we need to add a middleware loader + more to have each validation step as a middleware?

- Headers are now set correctly, as long as the configuration is a JSON string. Do you feel this is a good approach?

- Response code is now set correctly.

- I changed my context to include timeouts, currently its a hardcoded value, would it make sense to make this a configuration item? Same that the HTTP server currently runs with default HTTP timeouts as well, is this something I should add?

I was a bit unsure what you meant with overriding headers from ResponseBody configuration, is the current implementation in which I set the response headers before calling `sendResponse()` not sufficient? Do you have any examples on how you would like to see it?

P1llus commented:
@kvch Added first iteration of docs, let me know if it looks okay :) Do I need to add it anywhere else as well? I don't think it will be automatically included by just creating a new asciidoc in the inputs docs folder.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,279 @@
+	})
+	if !event {
+		in.sendResponse(w, http.StatusInternalServerError, in.createErrorMessage("Unable to send event"))
+		return
```
nit: this and the return outside of the block can be removed

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,279 @@
+	defer in.log.Infof("%v worker has stopped.", inputName)
+	in.server.Start()
+
+	return
```
this can be removed, just use the implicit `return`

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,279 @@
+		return http.StatusUnsupportedMediaType, in.createErrorMessage("Wrong Content-Type header, expecting application/json")
+	}
+
+	if r.Header.Get("Accept") != "application/json" {
```
So, my previous comment about overriding headers mainly stems from this line. The `Accept` header checking seems to indicate that we're always going to respond with a `Content-Type: application/json` header--but the fact that we allow the response headers to be overridden in the config actually breaks this. Nothing currently prevents the user from doing something like:

```
{beatname_lc}.inputs:
- type: http_endpoint
  ...
  response_body: '<message>success</message>'
  response_header: '{"Content-Type": "application/xml"}'
```

If we're actually wanting to make sure we always hand back `json` I'd do two things:

1. Always set the `Content-Type: application/json` header and ignore what the user may specify in the config
2. Validate the response value that the user specifies in `response_body` is actually valid JSON.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/input.go:
```diff
@@ -0,0 +1,279 @@
+// Triggers if middleware validation returns successful
+func (in *HttpEndpoint) apiResponse(w http.ResponseWriter, r *http.Request) {
+	objmap := make(map[string]string)
+	json.Unmarshal([]byte(in.config.ResponseHeader), &objmap)
```
I think it would make more sense to me to make `ResponseHeader` something like `map[string][]string` a la https://golang.org/pkg/net/http/#Header. That way there's no need to json-encode the header values in the configuration.

P1llus commented:
@andrewstucki 
I thought more about the response header for now, and it makes sense to me to remove it for now.
There is a few reasons for that.
1. If we decide later on to add support for more than JSON, it would make more sense.
2. Currently the custom response headers was more to add other headers than content-type, though I am still having a hard time finding good use-cases for when that would be needed.
So if it is okay with you, I have hardcoded the response header (upon success), but still allowing anyone to set the response code and will revisit this later if there is requests open for this feature.

andrewstucki commented on x-pack/filebeat/input/http_endpoint/config.go:
```diff
@@ -0,0 +1,37 @@
+	Username      string                  `config:"username"`
+	Password      string                  `config:"password"`
+	ResponseCode  int                     `config:"response_code" validate:"positive"`
+	ResponseBody  string                  `config:"response_body"`
```
If we're going to hardcode that this is a JSON-based response, I think we should likely still sanity check that the user actually entered JSON. You could just make a `Validate` method for the `config` structure and inside do something like 
```go
func (c *config) Validate() error {
  if !json.Valid(c.ResponseBody) {
    return errors.New("response_body must be valid JSON")
  }
  // other validations here
  return nil 
}
```

P1llus replied:
Added the validation, just had to convert it from a string to a byteslice as well.

New return upon wrong response body is:
```
2020-05-25T08:43:01.744+0200	ERROR	instance/beat.go:931	Exiting: Failed to start crawler: starting input failed: Error while initializing input: response_body must be valid JSON accessing 'filebeat.inputs.0' (source:'filebeat.yml')
Exiting: Failed to start crawler: starting input failed: Error while initializing input: response_body must be valid JSON accessing 'filebeat.inputs.0' (source:'filebeat.yml')
```

P1llus commented:
Should now be ready for merge and backport :)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/25//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/25//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/25//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18298/detail/PR-18298/25//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_18298.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #18298 updated]`

* Start Time: 2020-05-25T06:45:34.618+0000

* Duration: 17 min 52 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Make check`
    * Description: make check
    
    * Duration: 10 min 57 sec
    
    * Start Time: 2020-05-25T06:52:01.862+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats-beats-mbp/pipelines/PR-18298/runs/25/steps/324/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-05-25T07:02:03.130Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.131Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.131Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-25T07:02:03.132Z] Stage "Kubernetes" skipped due to earlier failure(s)
[2020-05-25T07:02:03.226Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.228Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-25T07:02:03.229Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.230Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-05-25T07:02:03.231Z] Stage "Packetbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.232Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.233Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.234Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.235Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.236Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-25T07:02:03.611Z] Failed in branch Elastic Agent x-pack
[2020-05-25T07:02:03.613Z] Failed in branch Elastic Agent x-pack Windows
[2020-05-25T07:02:03.614Z] Failed in branch Elastic Agent Mac OS X
[2020-05-25T07:02:03.614Z] Failed in branch Filebeat oss
[2020-05-25T07:02:03.615Z] Failed in branch Filebeat x-pack
[2020-05-25T07:02:03.616Z] Failed in branch Filebeat Mac OS X
[2020-05-25T07:02:03.617Z] Failed in branch Filebeat x-pack Mac OS X
[2020-05-25T07:02:03.618Z] Failed in branch Filebeat Windows
[2020-05-25T07:02:03.618Z] Failed in branch Filebeat x-pack Windows
[2020-05-25T07:02:03.619Z] Failed in branch Auditbeat x-pack
[2020-05-25T07:02:03.620Z] Failed in branch Libbeat x-pack
[2020-05-25T07:02:03.621Z] Failed in branch Metricbeat OSS Unit tests
[2020-05-25T07:02:03.622Z] Failed in branch Metricbeat OSS Integration tests
[2020-05-25T07:02:03.623Z] Failed in branch Metricbeat Python integration tests
[2020-05-25T07:02:03.623Z] Failed in branch Metricbeat crosscompile
[2020-05-25T07:02:03.624Z] Failed in branch Metricbeat Mac OS X
[2020-05-25T07:02:03.625Z] Failed in branch Metricbeat x-pack Mac OS X
[2020-05-25T07:02:03.626Z] Failed in branch Metricbeat Windows
[2020-05-25T07:02:03.627Z] Failed in branch Metricbeat x-pack Windows
[2020-05-25T07:02:03.628Z] Failed in branch Winlogbeat Windows x-pack
[2020-05-25T07:02:03.629Z] Failed in branch Kubernetes
[2020-05-25T07:02:03.904Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.906Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-25T07:02:03.909Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.910Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-05-25T07:02:03.911Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.912Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:03.913Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-25T07:02:03.970Z] Failed in branch Packetbeat
[2020-05-25T07:02:03.972Z] Failed in branch dockerlogbeat
[2020-05-25T07:02:03.973Z] Failed in branch Journalbeat
[2020-05-25T07:02:04.192Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:04.194Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-25T07:02:04.195Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:04.196Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-05-25T07:02:04.197Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-25T07:02:04.232Z] Failed in branch Metricbeat x-pack
[2020-05-25T07:02:04.233Z] Failed in branch Winlogbeat
[2020-05-25T07:02:04.428Z] Failed in branch Heartbeat
[2020-05-25T07:02:04.428Z] Failed in branch Libbeat
[2020-05-25T07:02:04.429Z] Failed in branch Functionbeat
[2020-05-25T07:02:04.429Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-05-25T07:02:04.431Z] Stage "Generators" skipped due to earlier failure(s)
[2020-05-25T07:02:04.534Z] Failed in branch Auditbeat oss
[2020-05-25T07:02:04.535Z] Failed in branch Generators
[2020-05-25T07:02:04.657Z] Running in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-25T07:02:04.970Z] + find . -type f -name TEST*.xml -path */build/* -delete
[2020-05-25T07:02:04.986Z] Running in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats/Lint
[2020-05-25T07:02:05.366Z] + cat
[2020-05-25T07:02:05.366Z] + /usr/local/bin/runbld ./runbld-script
[2020-05-25T07:02:05.366Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-05-25T07:02:11.994Z] runbld>>> runbld started
[2020-05-25T07:02:11.994Z] runbld>>> 1.6.11/a66728ff8f4356963772e6e6d2069392fa06acbe
[2020-05-25T07:02:13.924Z] runbld>>> The following profiles matched the job 'Beats/beats-beats-mbp/PR-18298' in order of occurrence in the config (last value wins).
[2020-05-25T07:02:15.323Z] runbld>>> Debug logging enabled.
[2020-05-25T07:02:15.587Z] runbld>>> Storing result
[2020-05-25T07:02:15.587Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-05-25T07:02:15.587Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1587637540455/t/20200525070215-6D03CCCD
[2020-05-25T07:02:15.587Z] runbld>>> Adding system facts.
[2020-05-25T07:02:16.548Z] runbld>>> Adding vcs info for the latest commit:  f1ab689e27a3adcd8bbc9fc32b39ad52f3304c33
[2020-05-25T07:02:16.812Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-05-25T07:02:16.812Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-05-25T07:02:16.812Z] + echo 'Processing JUnit reports with runbld...'
[2020-05-25T07:02:16.812Z] Processing JUnit reports with runbld...
[2020-05-25T07:02:17.389Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-05-25T07:02:17.389Z] runbld>>> DURATION: 20ms
[2020-05-25T07:02:17.389Z] runbld>>> STDOUT: 40 bytes
[2020-05-25T07:02:17.389Z] runbld>>> STDERR: 49 bytes
[2020-05-25T07:02:17.389Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-05-25T07:02:17.389Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-25T07:02:18.340Z] runbld>>> Storing build metadata: 
[2020-05-25T07:02:18.340Z] runbld>>> Adding test report.
[2020-05-25T07:02:18.340Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298/src/github.com/elastic/beats
[2020-05-25T07:02:19.289Z] runbld>>> Found 0 test output files
[2020-05-25T07:02:19.290Z] runbld>>> Test output logs contained: Errors: 0 Failures: 0 Tests: 0 Skipped: 0
[2020-05-25T07:02:19.290Z] runbld>>> Storing result
[2020-05-25T07:02:19.553Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-05-25T07:02:19.553Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1587637540455/t/20200525070215-6D03CCCD
[2020-05-25T07:02:19.553Z] runbld>>> Email notification disabled by environment variable.
[2020-05-25T07:02:19.553Z] runbld>>> Slack notification disabled by environment variable.
[2020-05-25T07:02:25.833Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-18298
[2020-05-25T07:02:26.143Z] [INFO] getVaultSecret: Getting secrets
[2020-05-25T07:02:26.194Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-05-25T07:02:26.860Z] + chmod 755 generate-build-data.sh
[2020-05-25T07:02:26.860Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/runs/25 FAILURE 1011972
[2020-05-25T07:02:27.411Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-18298/runs/25/steps/?limit=10000 -o steps-info.json

```

</p>
</details>

P1llus commented:
jenkins, test this please

kvch commented:
jenkins run tests
